spec/mempool/mempool.md

```diff
@@ -0,0 +1,152 @@
+
+## Interactions
+
+In what follows, we present the interactions of the mempool with other parts of the CometBFT protocol.
+Some of the specificities of the current implementation (`CListMempool`) are also detailed.
+
+**RPC server**
+To add a new transaction to the mempool, a clients may submit it through an appropriate RPC endpoint.
+This endpoint is offered by some of the system nodes (but not necessarily all of them).
+
+**Gossip protocol** 
+Transactions can also be received from other nodes, through a gossiping mechanism.
+
+**ABCI application**
+As pointed above, the mempool should only store and disseminate  valid transactions.
+It is up to the ABCI (client) application to define whether a transaction is valid.
+Transactions received locally are sent to the application to be validated, through the `checkTx` method from the mempool ABCI connection.
+Such a check indicates with a flag whether it is the first time (or not) that the transaction is received.
+Transactions that are validated by the application are later added to the mempool.
+Transactions tagged as invalid are simply drooped.
+The validity of a transaction may depend on the state of the client application.
+In particular, some transactions that are valid in some state of the application may later become invalid.
+The state of the application is updated when consensus commits a block of transactions.
+When this happens, the transactions still in the mempool have to be validated again.
+We further detail this mechanism below.
+
+**Consensus**
+The consensus protocol consumes transactions stored in the mempool to build blocks to be proposed.
+To this end, consensus requests from the mempool a list of transactions which abide by certain limits (namely, total number of transactions included, or total size in bytes).
+In the current implementation, the mempool is a list of transactions.
+Such a call returns the longest prefix of the list that is matching the requirements.
+Notice that at this point the transactions returned to consensus are not removed from the mempool.
+This comes from the fact that the block is proposed but not decided yet.
+
+Proposing a block is the prerogative of the nodes acting as validators.
+At all the nodes (validators or not), consensus is also responsible for committing blocks of transactions to the blockchain.
+Once a block is committed, all the transactions included in the block are removed from the mempool.
+This happens with an `update` call to the mempool.
+Before doing this call, consensus takes a `lock` on the mempool.
+It then `flush` the connection with the client application.
+Both operations aim at preventing any concurrent `checkTx` while the mempool is updated.
+At the end of `update`, all the transactions still in the mempool are re-validated against the new state of the client application.
+This procedure is executed asynchronously with a call to `recheckTxs`.
```

Comment:
To be more precise, sending re-CheckTx requests with `recheckTxs` happens at the end of `Update`. Processing the responses happens asynchronously.